File: python/examples/egc/json2egc.py
#!/usr/bin/env python3
"""
Parse EGC data from JSON and output its content as the here-defined EGC format

Usage: ./json2egc.py <inputfile> <inputspec>

Arguments:
  <inputfile>  Input file in Json format
  <inputspec>  Specification file describing the EGC format
"""
from textformats import Specification
from docopt import docopt
import json

args = docopt(__doc__)
with open(args["<inputfile>"]) as f:
  print(Specification(args["<inputspec>"])["file"].encode(f.read(),
    from_json=True))

File: examples/egc/json2egc.nim
import os, json, strutils
import textformats

const HelpMsg = """
Parse EGC data from JSON and output its content in EGC format

Usage: $# <inputfile> <inputspec>

Arguments:
  <inputfile>  Input file in Json format
  <inputspec>  Specification file describing the EGC format
"""

proc parse_args(): (string, Specification) =
  if (paramCount() != 2):
    echo(HelpMsg % [getAppFilename()])
    quit(0)
  result[0] = paramStr(1)
  result[1] = specification_from_file(paramStr(2))

when isMainModule:
  try:
    let
      (input_file, spec) = parse_args()
      ddef = get_definition(spec, "file")
      input_data = parse_json(read_file(input_file))
    echo(input_data.encode(ddef))
    quit(0)
  except:
    echo(getCurrentExceptionMsg())
    quit(1)
